pyloidal/cocos.py

```diff
@@ -0,0 +1,221 @@
+r"""
+Determine the COCOS (COordinate COnventionS) convention used to describe a tokamak
+equilibrium and convert between COCOS systems. Based on the work of Sauter, O. and
+Medvedev, S.Y., 2013. Tokamak coordinate conventions: COCOS. *Computer Physics
+Communications*, 184(2), pp.293-302.
+
+Throughout, we denote the coordinate systems in a tokamak with the following terms:
+
+- :math:`R`, the major radius
+- :math:`r`, the minor radius
+- :math:`Z`, the vertical coordinate
+- :math:`\phi`, the toroidal angle
+- :math:`\theta`, the poloidal angle
+
+These functions were adapted from OMAS (Copyright MIT License, 2017, Orso Meneghini).
+"""
+
+import itertools
+from typing import Dict, Optional, Tuple, Union
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+def sigma_to_cocos(
+    sigma_bp: int,
+    sigma_rpz: int,
+    sigma_rtp: int,
+    psi_by_2pi: bool = True,
+) -> int:
+    r"""
+    We can (partially) determine the COCOS by checking the :math:`\sigma` quantities,
+    defined below. We additionally need to know whether :math:`\psi` is defined with a
+    factor of :math:`1/(2\pi)` to determine whether the COCOS is in the range 1 to 8 or
+    11 to 18.
+
+    Parameters
+    ----------
+    sigma_rtp
+        :math:`\sigma_{B_p}`: Given by :math:`\text{sign}(\vec{B}_p`\cdot\nabla\theta)`,
+        so is +1 when the poloidal magnetic field is in the same direction as increasing
+        :math:`\theta` and -1 when they are opposed.
+    sigma_rpz
+        :math:`\sigma_{R \phi Z`}: +1 when :math:`(R, \phi`, Z)` form a right-handed
+        coordinate system, and -1 when :math:`(R, Z, \phi)` form a right-handed
+        coordinate system.
+    sigma_bp
+        :math:`\sigma_{r\theta\phi}`: +1 when :math:`(r, \theta, \phi)` form a
+        right-handed coordinate system, and -1 when :math:`(r, \theta, \phi)` form a
+        right-handed coordainte system.
+    psi_by_2pi
+        If true, :math:`\psi` is defined with a factor of :math:`1/(2\pi)`.
+
+    Returns
+    -------
+    int
+        COCOS convention in use.
+    """
+    if sigma_bp != 1 and sigma_bp != -1:
+        raise ValueError(f"sigma_bp should be either 1 or -1, found {sigma_bp}")
+    if sigma_rpz != 1 and sigma_rpz != -1:
+        raise ValueError(f"sigma_rpz should be either 1 or -1, found {sigma_rpz}")
+    if sigma_rtp != 1 and sigma_rtp != -1:
+        raise ValueError(f"sigma_rtp should be either 1 or -1, found {sigma_rtp}")
+
+    sigma_to_cocos_dict = {
+        (+1, +1, +1): 1,  # +Bp, +rpz, +rtp
```

Comment:
Future change: move this dict out of this function, and make keys a frozen dataclass?

Reply:
Agreed, I'd originally written a `NamedTuple` that achieved a similar thing to what you described, but was spending far too long getting everything documented properly and in the end reverted to the current method. I was also thinking of explicitly defining the entire standard COCOS table, and functions could then just perform lookups on it.